src/c/device-coap.c

```diff
@@ -99,50 +103,130 @@ static bool coap_init
   return true;
 }
 
-static bool coap_get_handler
-(
-  void *impl,
-  const devsdk_device_t *device,
-  uint32_t nreadings,
-  const devsdk_commandrequest *requests,
-  devsdk_commandresult *readings,
-  const iot_data_t *options,
-  iot_data_t **exception
-)
-{
-  (void) impl;
-  (void) device;
-  (void) nreadings;
-  (void) requests;
-  (void) readings;
-  (void) options;
-
-  *exception = iot_data_alloc_string (NOT_SUPPORTED_TEXT, IOT_DATA_REF);
-  return false;
+static bool coap_get_handler(void *impl, const devsdk_device_t *device,
+                             uint32_t nreadings,
+                             const devsdk_commandrequest *requests,
+                             devsdk_commandresult *readings,
+                             const iot_data_t *options,
+                             iot_data_t **exception) {
+  bool successful_get_request = true;
+  coap_driver *driver = (coap_driver *)impl;
+  int ret = EXIT_FAILURE;
+  uint32_t i = 0;
+
+  if (device == NULL) {
+    iot_log_error(driver->lc, "COAP:Device is empty");
+    return successful_get_request;
+  }
+  successful_get_request = GetEndDeviceConfig(device->name);
```

Comment:
`GetEndDeviceConfig` works by looking up the named device and parsing its protocol properties. Instead, look at `device->address`. As the code stands, this is the `devsdk_protocols` object that you need. A further refinement would be to enhance `coap_create_address` and `coap_free_address` to translate the protocols into a parsed structure of your choice. This is then only performed once per device, rather than on every call to get/set.

Reply:
Removed `GetEndDeviceConfig` function, and updating end device properties using `coap_create_address` as suggested.  Added pthread mutex for synchronization between multiple threads. As `coap_resp_data` a global variable getting updated in coap call back handler and used in `coap_get_handler`.  As this sharing of global variable by multiple threads leads to race condition, mutex is used to avoid the same. Tried different methods to fix the race condition, using mutex is working as expected. This synchronization method is referred from another edgex device service called `device-grove-c` . Let me know your thoughts on this addition.